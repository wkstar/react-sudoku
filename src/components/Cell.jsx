import React, { useState } from 'react';
import styles from './Cell.module.css';

export default function Cell(props) {
  const { cell } = props;
  const [answer, setAnswer] = useState(() => {
    if (cell.getRevealed()) {
      return cell.getSolution();
    }
  });
  const [correct, setCorrect] = useState();

  const handleAnswerChange = event => {
    const newAnswer = event.target.value;
    const isCorrectFormat = /^[0-9]$|^$/g.test(newAnswer);
    if (isCorrectFormat) {
      setAnswer(newAnswer);
    }
    const isCorrect = cell.getSolution().toString() === newAnswer || newAnswer === '';
    setCorrect(isCorrect);
  };

  const cellInputWrongClass = (correct === false) ? styles.cell_input_wrong : '';
  return (
    <div className={styles.cell}>
      <input className={`${styles.cell_input} ${cellInputWrongClass}`} value={answer} onChange={handleAnswerChange} />
    </div>
  );
}
