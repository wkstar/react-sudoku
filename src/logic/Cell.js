const GRID_SIZE = 9;
const VALID_ANSWERS = [1, 2, 3, 4, 5, 6, 7, 8, 9];

export default class SudukoCell {
  constructor(index) {
    this.solution = null;
    this.index = index;
    this.coords = this.getCoords(index);
    this.neighbours = this.getNeighbours();
  }

  getSolution() {
    return this.solution;
  }

  isValidSolution(proposeSolution) {
    const neighbourSolutions = this.getNeighbourSolutions();
    return neighbourSolutions.contains(proposedSolution);
  }

  getNeighbourSolutions() {
    return this.neighbours.map(neighbour => {
      return neighbour.getSolution();
    });
  }

  getNeighbours() {
    const boxNeighbours = getBoxNeighbours();
    const rowNeighbours = getRowNeighbours();
    const columnNeighbours = getColumnNeighbours();
    return unique(boxNeighbours + rowNeighbours + columnNeighbours);
  }

  getCoords(index) {}
}
