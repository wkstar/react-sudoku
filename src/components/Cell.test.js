import React from 'react';
import { shallow } from 'enzyme';
import Cell from './Cell';
import TestRenderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';

const mockCell = (solution, revealed) => ({
  getSolution: () => solution,
  getRevealed: () => revealed
});
describe('Cell', () => {
  it('should render a revealed cell', () => {
    const wrapper = shallow(<Cell cell={mockCell(123, true)} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should render a hidden cell', () => {
    const wrapper = shallow(<Cell cell={mockCell(321, false)} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should add an error class when the wrong answer is given', () => {
    // This works, but because react hooks are bleeding edge and enzyme hasn't caught up yet it throws some warnings.
    let testRenderer;
    let input;
    act(() => {
      testRenderer = TestRenderer.create(<Cell cell={mockCell(8, false)} />);
      const testInstance = testRenderer.root;
      input = testInstance.findByType('input');
    });

    act(() => {
      input.props['onChange']({
        target: {
          value: 9
        }
      });
    });

    expect(input.props['className']).toBe('cell_input cell_input_wrong');
  });
});
