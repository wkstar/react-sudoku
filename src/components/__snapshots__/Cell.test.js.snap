// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell should render a hidden cell 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Cell
    cell={
      Object {
        "getRevealed": [Function],
        "getSolution": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <input
        className="cell_input "
        onChange={[Function]}
      />,
      "className": "cell",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "cell_input ",
        "onChange": [Function],
        "value": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": "input",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <input
          className="cell_input "
          onChange={[Function]}
        />,
        "className": "cell",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "cell_input ",
          "onChange": [Function],
          "value": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Cell should render a revealed cell 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Cell
    cell={
      Object {
        "getRevealed": [Function],
        "getSolution": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <input
        className="cell_input "
        onChange={[Function]}
        value={123}
      />,
      "className": "cell",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "cell_input ",
        "onChange": [Function],
        "value": 123,
      },
      "ref": null,
      "rendered": null,
      "type": "input",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <input
          className="cell_input "
          onChange={[Function]}
          value={123}
        />,
        "className": "cell",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "cell_input ",
          "onChange": [Function],
          "value": 123,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
